File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t
  def verse(number) do
    bottles_of_beer_lyrics(number) <> take_one_down_lyrics(number)
  end

  defp bottles_of_beer_lyrics(verse_number) do
    bottle_count = verse_number - 1
    sentence_case("#{beer_on_the_wall(bottle_count)}, #{pluralize("bottle", bottle_count)} of beer.\n")
  end

  defp pluralize(word, count) do
    case count do
      0 -> "no more #{word}s"
      1 -> "1 #{word}"
      _ -> "#{count} #{word}s"
    end
  end

  defp sentence_case(<< first_letter::binary-size(1) >> <> rest) do
    String.upcase(first_letter) <> rest
  end

  defp beer_on_the_wall(bottle_count) do
    "#{pluralize("bottle", bottle_count)} of beer on the wall"
  end

  defp take_one_down_lyrics(number) do
    case number do
      1 -> "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
      2 -> "Take it down and pass it around, no more bottles of beer on the wall.\n"
      _ -> "Take one down and pass it around, #{beer_on_the_wall(number - 2)}.\n"
    end
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t) :: String.t
  def lyrics(range) do
    range
    |> Enum.map(&verse/1)
    |> Enum.join("\n")
  end

  def lyrics, do: lyrics(100..1)

end

File: java/beer-song/src/main/java/BeerSong.java
import java.util.ArrayList;
import java.util.List;

public class BeerSong {

  public static String verse(int number) {
    return singBottlesOfBeer(number) + singTakeOneDown(number);
  }

  private static String singBottlesOfBeer(int number) {
    return capitalize(singBeerOnWall(number) + ", " + pluralize("bottle", number) + " of beer.\n");
  }

  private static String singBeerOnWall(int count) {
    return String.format("%s of beer on the wall", pluralize("bottle", count));
  }

  public static String sing(int start, int end) {
    List<String> verses = new ArrayList<>();

    for (int verseNumber = start; verseNumber >= end; --verseNumber) {
      verses.add(verse(verseNumber));
    }

    return String.join("", verses);
  }

  public static String singSong() {
    return sing(99, 0);
  }

  private static String singTakeOneDown(int number) {
    switch (number) {
      case 0:
        return "Go to the store and buy some more, 99 bottles of beer on the wall.\n\n";
      case 1:
        return "Take it down and pass it around, no more bottles of beer on the wall.\n\n";
      default:
        return String.format("Take one down and pass it around, %s.\n\n", singBeerOnWall(number - 1));
    }
  }

  private static String capitalize(String sentence) {
    return sentence.substring(0, 1).toUpperCase() + sentence.substring(1);
  }

  private static String pluralize(String word, int count) {
    switch (count) {
      case 0:
        return String.format("no more %ss", word);
      case 1:
        return String.format("1 %s", word);
      default:
        return String.format("%d %ss", count, word);
    }
  }

}